062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+The `goals` parameter can be used to specify a comma-separated list of goals used to generate the proposal for adding or removing replicas (default goals are used when not set).
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
+
+There is also a `replication_throttle` parameter, which is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is already configurable in `Kafka.spec.cruiseControl.config`, and it also applies to all task executions, including replication factor changes.
+
+#### User tasks
+
+We can use the `user_tasks` endpoint to get the task execution result, passing the `User-Task-ID` header value returned by the `topic_configuration` response.
+
+Example:
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Corner cases
+
+Here we analyze some events that may happen during the replication factor change execution, and how they are handled by the Topic Operator.
+
+#### Topic deleted
+
+If it's deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, and logs an error for the ongoing replication factor change, unless the `KafkaTopic` is paused or unmanaged.
+If it's deleted in Kafka, the Topic Operator recreates the topic in Kafka with the target replication factor, unless the `KafkaTopic` is paused or unmanaged.
+
+#### Cruise Control crashes
+
+Pending replication factor change requests are not durable, which means they are lost if Cruise Control crashes or is restarted after accepting the request.
+In that case, the Topic Operator gets an exception while polling, removes the pending change from its in-memory cache, and retries until Cruise Control comes back.
+
+#### Multiple changes for the same topic
+
+A reconciliation batch can only contain one event related to a specific topic, other events for the same topic are pushed back to the from of the queue.
+It is still possible that one or more replication factor changes for the same topic are requested while another one is already ongoing.
+In this case, the Topic Operator ignores intermediate changes, and only reconciles the last update when the ongoing change in completed.
+
+#### Topic Operator crashes
+
+In this case, there may be some pending and some ongoing replication factor changes.
+All pending and ongoing changes are recovered by the periodic reconciliation as soon as the Topic Operator restarts.
```

Comment:
How does the Topic Operator(TO) recover the pending and ongoing changes? I assume the TO uses the `generation` and `observedGeneration` fields of the `KafkaTopic` resource to find which topics still need to be updated, then uses the information from those `KafkaTopic` resources to query the CC API to find which have ongoing tasks? Then adding any topics without any ongoing tasks to TO's in-memory cache for future CC task queuing?

Reply:
The `observedGeneration` means that the operator saw this change. So that should be updated right after the operator sees the change and not be used to track some in-progress rebalancing.

Reply:
Yes, I'm doing as Jakub suggests. 

When recovering after a TO restart, the reconciliation logic is used to rebuild the in-memory cache. Both pending and ongoing changes are detected by comparing with the actual topic RF. The only difference is that ongoing changes are effectively duplicates, because they are already known by Cruise Control. This means that, if the task is still `Active` or `InExecution`, we may see a `Warning` with the message "Another task is executing, will retry later", but then it will still converge to `Ready` state as soon as the original execution completes.